File: app/src/main/java/com/nextcloud/talk/controllers/bottomsheet/EntryMenuController.java
/*
 * Nextcloud Talk application
 *
 * @author Mario Danic
 * Copyright (C) 2017 Mario Danic <mario@lovelyhq.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.nextcloud.talk.controllers.bottomsheet;

import android.content.ComponentName;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.ImageView;

import com.bluelinelabs.conductor.RouterTransaction;
import com.bluelinelabs.conductor.changehandler.HorizontalChangeHandler;
import com.google.android.material.textfield.TextInputLayout;
import com.nextcloud.talk.R;
import com.nextcloud.talk.application.NextcloudTalkApplication;
import com.nextcloud.talk.controllers.base.BaseController;
import com.nextcloud.talk.models.json.conversations.Conversation;
import com.nextcloud.talk.utils.EmojiTextInputEditText;
import com.nextcloud.talk.utils.ShareUtils;
import com.nextcloud.talk.utils.bundle.BundleKeys;
import com.nextcloud.talk.utils.database.user.UserUtils;
import com.nextcloud.talk.utils.singletons.ApplicationWideMessageHolder;
import com.vanniktech.emoji.EmojiImageView;
import com.vanniktech.emoji.EmojiPopup;
import com.vanniktech.emoji.emoji.Emoji;
import com.vanniktech.emoji.listeners.OnEmojiClickListener;
import com.vanniktech.emoji.listeners.OnEmojiPopupDismissListener;
import com.vanniktech.emoji.listeners.OnEmojiPopupShownListener;

import org.greenrobot.eventbus.EventBus;
import org.parceler.Parcels;

import javax.inject.Inject;

import androidx.annotation.NonNull;
import autodagger.AutoInjector;
import butterknife.BindView;
import butterknife.OnClick;

@AutoInjector(NextcloudTalkApplication.class)
public class EntryMenuController extends BaseController {

    @BindView(R.id.ok_button)
    Button proceedButton;

    @BindView(R.id.text_edit)
    EmojiTextInputEditText editText;

    @BindView(R.id.text_input_layout)
    TextInputLayout textInputLayout;

    @BindView(R.id.smileyButton)
    ImageView smileyButton;

    @Inject
    EventBus eventBus;

    @Inject
    UserUtils userUtils;

    private ConversationOperationEnum operation;
    private Conversation conversation;
    private Intent shareIntent;
    private String packageName;
    private String name;
    private String callUrl;

    private EmojiPopup emojiPopup;

    private Bundle originalBundle;

    public EntryMenuController(Bundle args) {
        super(args);
        originalBundle = args;

        this.operation = (ConversationOperationEnum) args.getSerializable(BundleKeys.INSTANCE.getKEY_OPERATION_CODE());
        if (args.containsKey(BundleKeys.INSTANCE.getKEY_ROOM())) {
            this.conversation = Parcels.unwrap(args.getParcelable(BundleKeys.INSTANCE.getKEY_ROOM()));
        }

        if (args.containsKey(BundleKeys.INSTANCE.getKEY_SHARE_INTENT())) {
            this.shareIntent = Parcels.unwrap(args.getParcelable(BundleKeys.INSTANCE.getKEY_SHARE_INTENT()));
        }

        this.name = args.getString(BundleKeys.INSTANCE.getKEY_APP_ITEM_NAME(), "");
        this.packageName = args.getString(BundleKeys.INSTANCE.getKEY_APP_ITEM_PACKAGE_NAME(), "");
        this.callUrl = args.getString(BundleKeys.INSTANCE.getKEY_CALL_URL(), "");
    }

    @NonNull
    @Override
    protected View inflateView(@NonNull LayoutInflater inflater, @NonNull ViewGroup container) {
        return inflater.inflate(R.layout.controller_entry_menu, container, false);
    }

    @OnClick(R.id.smileyButton)
    void onSmileyClick() {
        emojiPopup.toggle();
    }

    @Override
    protected void onAttach(@NonNull View view) {
        super.onAttach(view);
        if (ApplicationWideMessageHolder.getInstance().getMessageType() != null &&
                ApplicationWideMessageHolder.getInstance().getMessageType().equals(ApplicationWideMessageHolder.MessageType.CALL_PASSWORD_WRONG)) {
            textInputLayout.setError(getResources().getString(R.string.nc_wrong_password));
            ApplicationWideMessageHolder.getInstance().setMessageType(null);
            if (proceedButton.isEnabled()) {
                proceedButton.setEnabled(false);
                proceedButton.setAlpha(0.7f);
            }
        }
    }

    @OnClick(R.id.ok_button)
    public void onProceedButtonClick() {
        Bundle bundle;
        if (operation == ConversationOperationEnum.JOIN_ROOM) {
            bundle = new Bundle();
            bundle.putParcelable(BundleKeys.INSTANCE.getKEY_ROOM(), Parcels.wrap(conversation));
            bundle.putString(BundleKeys.INSTANCE.getKEY_CALL_URL(), callUrl);
            bundle.putString(BundleKeys.INSTANCE.getKEY_CONVERSATION_PASSWORD(), editText.getText().toString());
            bundle.putSerializable(BundleKeys.INSTANCE.getKEY_OPERATION_CODE(), operation);
            if (originalBundle.containsKey(BundleKeys.INSTANCE.getKEY_SERVER_CAPABILITIES())) {
                bundle.putParcelable(BundleKeys.INSTANCE.getKEY_SERVER_CAPABILITIES(), originalBundle.getParcelable(BundleKeys.INSTANCE.getKEY_SERVER_CAPABILITIES()));
            }

            getRouter().pushController(RouterTransaction.with(new OperationsMenuController(bundle))
                    .pushChangeHandler(new HorizontalChangeHandler())
                    .popChangeHandler(new HorizontalChangeHandler()));
        } else if (operation != ConversationOperationEnum.SHARE_LINK && operation != ConversationOperationEnum.GET_JOIN_ROOM && operation != ConversationOperationEnum.INVITE_USERS) {
            bundle = new Bundle();
            if (operation == ConversationOperationEnum.CHANGE_PASSWORD || operation == ConversationOperationEnum.SET_PASSWORD) {
                conversation.setPassword(editText.getText().toString());
            } else {
                conversation.setName(editText.getText().toString());
            }
            bundle.putParcelable(BundleKeys.INSTANCE.getKEY_ROOM(), Parcels.wrap(conversation));
            bundle.putSerializable(BundleKeys.INSTANCE.getKEY_OPERATION_CODE(), operation);
            getRouter().pushController(RouterTransaction.with(new OperationsMenuController(bundle))
                    .pushChangeHandler(new HorizontalChangeHandler())
                    .popChangeHandler(new HorizontalChangeHandler()));
        } else if (operation == ConversationOperationEnum.SHARE_LINK) {
            if (getActivity() != null) {
                shareIntent.putExtra(Intent.EXTRA_TEXT, ShareUtils.getStringForIntent(getActivity(),
                        editText.getText().toString(), userUtils, conversation));
                Intent intent = new Intent(shareIntent);
                intent.setComponent(new ComponentName(packageName, name));
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                getActivity().startActivity(intent);
            }
        } else if (operation != ConversationOperationEnum.INVITE_USERS) {
            bundle = new Bundle();
            bundle.putSerializable(BundleKeys.INSTANCE.getKEY_OPERATION_CODE(), operation);
            bundle.putString(BundleKeys.INSTANCE.getKEY_CALL_URL(), editText.getText().toString());
            getRouter().pushController(RouterTransaction.with(new OperationsMenuController(bundle))
                    .pushChangeHandler(new HorizontalChangeHandler())
                    .popChangeHandler(new HorizontalChangeHandler()));

        } else if (operation == ConversationOperationEnum.INVITE_USERS) {
            originalBundle.putString(BundleKeys.INSTANCE.getKEY_CONVERSATION_NAME(), editText.getText().toString());
            getRouter().pushController(RouterTransaction.with(new OperationsMenuController(originalBundle))
                    .pushChangeHandler(new HorizontalChangeHandler())
                    .popChangeHandler(new HorizontalChangeHandler()));

        }
    }

    @Override
    protected void onViewBound(@NonNull View view) {
        super.onViewBound(view);
        NextcloudTalkApplication.Companion.getSharedApplication().getComponentApplication().inject(this);

        if (conversation != null && operation == ConversationOperationEnum.RENAME_ROOM) {
            editText.setText(conversation.getName());
        }

        editText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE && proceedButton != null && proceedButton.isEnabled()) {
                proceedButton.callOnClick();
                return true;
            }
            return false;
        });

        editText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (!TextUtils.isEmpty(s)) {
                    if (operation == ConversationOperationEnum.RENAME_ROOM) {
                        if (conversation.getName() == null || !conversation.getName().equals(s.toString())) {
                            if (!proceedButton.isEnabled()) {
                                proceedButton.setEnabled(true);
                                proceedButton.setAlpha(1.0f);
                            }
                            textInputLayout.setErrorEnabled(false);
                        } else {
                            if (proceedButton.isEnabled()) {
                                proceedButton.setEnabled(false);
                                proceedButton.setAlpha(0.38f);
                            }
                            textInputLayout.setError(getResources().getString(R.string.nc_call_name_is_same));
                        }
                    } else if (operation != ConversationOperationEnum.GET_JOIN_ROOM) {
                        if (!proceedButton.isEnabled()) {
                            proceedButton.setEnabled(true);
                            proceedButton.setAlpha(1.0f);
                        }
                        textInputLayout.setErrorEnabled(false);
                    } else if ((editText.getText().toString().startsWith("http://") ||
                            editText.getText().toString().startsWith("https://")) &&
                                    editText.getText().toString().contains("/call/")) {
                        // operation code 10
                        if (!proceedButton.isEnabled()) {
                            proceedButton.setEnabled(true);
                            proceedButton.setAlpha(1.0f);
                        }
                        textInputLayout.setErrorEnabled(false);
                    } else {
                        if (proceedButton.isEnabled()) {
                            proceedButton.setEnabled(false);
                            proceedButton.setAlpha(0.38f);
                        }
                        textInputLayout.setError(getResources().getString(R.string.nc_wrong_link));
                    }
                } else {
                    if (proceedButton.isEnabled()) {
                        proceedButton.setEnabled(false);
                        proceedButton.setAlpha(0.38f);
                    }
                    textInputLayout.setErrorEnabled(false);
                }
            }
        });

        String labelText = "";
        switch (operation) {
            case INVITE_USERS:
            case RENAME_ROOM:
                labelText = getResources().getString(R.string.nc_call_name);
                editText.setInputType(InputType.TYPE_CLASS_TEXT);
                smileyButton.setVisibility(View.VISIBLE);
                emojiPopup = EmojiPopup.Builder.fromRootView(view).setOnEmojiPopupShownListener(new OnEmojiPopupShownListener() {
                    @Override
                    public void onEmojiPopupShown() {
                        if (getResources() != null) {
                            smileyButton.setColorFilter(getResources().getColor(R.color.colorPrimary),
                                    PorterDuff.Mode.SRC_IN);
                        }
                    }
                }).setOnEmojiPopupDismissListener(new OnEmojiPopupDismissListener() {
                    @Override
                    public void onEmojiPopupDismiss() {
                        if (smileyButton != null) {
                            smileyButton.setColorFilter(getResources().getColor(R.color.emoji_icons),
                                    PorterDuff.Mode.SRC_IN);
                        }
                    }
                }).setOnEmojiClickListener(new OnEmojiClickListener() {
                    @Override
                    public void onEmojiClick(@NonNull EmojiImageView emoji, @NonNull Emoji imageView) {
                        editText.getEditableText().append(" ");
                    }
                }).build(editText);

                break;
            case CHANGE_PASSWORD:
                labelText = getResources().getString(R.string.nc_new_password);
                editText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
                break;
            case SET_PASSWORD:
            case SHARE_LINK:
            case JOIN_ROOM:
                // 99 is joining a conversation via password
                labelText = getResources().getString(R.string.nc_password);
                editText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_PASSWORD);
                break;
            case GET_JOIN_ROOM:
                labelText = getResources().getString(R.string.nc_conversation_link);
                editText.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_URI);
                break;
            default:
                break;
        }

        if (operation == ConversationOperationEnum.JOIN_ROOM
            || operation == ConversationOperationEnum.CHANGE_PASSWORD
            || operation == ConversationOperationEnum.SET_PASSWORD
            || operation == ConversationOperationEnum.SHARE_LINK) {
            textInputLayout.setEndIconMode(TextInputLayout.END_ICON_PASSWORD_TOGGLE);
        } else {
            textInputLayout.setEndIconMode(TextInputLayout.END_ICON_NONE);
        }

        textInputLayout.setHint(labelText);
        textInputLayout.requestFocus();
    }

    @Override
    public AppBarLayoutType getAppBarLayoutType() {
        return AppBarLayoutType.SEARCH_BAR;
    }
}

File: app/src/main/java/com/nextcloud/talk/controllers/bottomsheet/EntryMenuController.kt
/*
 * Nextcloud Talk application
 *
 * @author Mario Danic
 * @author Marcel Hibbe
 * @author Andy Scherzinger
 * Copyright (C) 2017 Mario Danic <mario@lovelyhq.com>
 * Copyright (C) 2022 Marcel Hibbe <dev@mhibbe.de>
 * Copyright (C) 2022 Andy Scherzinger <info@andy-scherzinger.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.nextcloud.talk.controllers.bottomsheet

import android.content.res.ColorStateList
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.core.content.res.ResourcesCompat
import autodagger.AutoInjector
import com.bluelinelabs.conductor.RouterTransaction
import com.bluelinelabs.conductor.changehandler.HorizontalChangeHandler
import com.google.android.material.textfield.TextInputLayout
import com.nextcloud.android.common.ui.theme.utils.ColorRole
import com.nextcloud.talk.R
import com.nextcloud.talk.api.NcApi
import com.nextcloud.talk.application.NextcloudTalkApplication
import com.nextcloud.talk.application.NextcloudTalkApplication.Companion.sharedApplication
import com.nextcloud.talk.controllers.base.BaseController
import com.nextcloud.talk.controllers.util.viewBinding
import com.nextcloud.talk.data.user.model.User
import com.nextcloud.talk.databinding.ControllerEntryMenuBinding
import com.nextcloud.talk.models.json.conversations.Conversation
import com.nextcloud.talk.models.json.conversations.RoomOverall
import com.nextcloud.talk.users.UserManager
import com.nextcloud.talk.utils.ApiUtils
import com.nextcloud.talk.utils.UriUtils
import com.nextcloud.talk.utils.bundle.BundleKeys
import com.nextcloud.talk.utils.bundle.BundleKeys.KEY_ROOM_TOKEN
import com.nextcloud.talk.utils.singletons.ApplicationWideMessageHolder
import com.vanniktech.emoji.EmojiPopup
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import okhttp3.internal.immutableListOf
import org.greenrobot.eventbus.EventBus
import javax.inject.Inject

@AutoInjector(NextcloudTalkApplication::class)
class EntryMenuController(args: Bundle) :
    BaseController(
        R.layout.controller_entry_menu,
        args
    ) {
    private val binding: ControllerEntryMenuBinding? by viewBinding(ControllerEntryMenuBinding::bind)

    @Inject
    lateinit var ncApi: NcApi

    @Inject
    lateinit var eventBus: EventBus

    @Inject
    lateinit var userManager: UserManager

    private val operation: ConversationOperationEnum
    private var conversation: Conversation? = null
    private val packageName: String
    private val name: String

    private var emojiPopup: EmojiPopup? = null
    private val originalBundle: Bundle
    private var currentUser: User? = null
    private val roomToken: String

    override val appBarLayoutType: AppBarLayoutType
        get() = AppBarLayoutType.SEARCH_BAR

    override fun onAttach(view: View) {
        super.onAttach(view)
        if (ApplicationWideMessageHolder.MessageType.CALL_PASSWORD_WRONG ==
            ApplicationWideMessageHolder.getInstance().messageType
        ) {
            binding?.textInputLayout?.error = resources?.getString(R.string.nc_wrong_password)
            ApplicationWideMessageHolder.getInstance().messageType = null
            if (binding?.okButton?.isEnabled == true) {
                binding?.okButton?.isEnabled = false
                binding?.okButton?.alpha = OPACITY_BUTTON_DISABLED
            }
        }
    }

    override fun onViewBound(view: View) {
        super.onViewBound(view)

        currentUser = userManager.currentUser.blockingGet()

        if (operation == ConversationOperationEnum.OPS_CODE_GET_AND_JOIN_ROOM) {
            var labelText = ""
            labelText = resources!!.getString(R.string.nc_conversation_link)
            binding?.textEdit?.inputType = InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_URI

            textEditAddChangedListener()

            binding?.textInputLayout?.let { viewThemeUtils.material.colorTextInputLayout(it) }
            binding?.okButton?.let { viewThemeUtils.material.colorMaterialButtonText(it) }

            binding?.textInputLayout?.hint = labelText
            binding?.textInputLayout?.requestFocus()

            binding?.smileyButton?.setOnClickListener { onSmileyClick() }
            binding?.okButton?.setOnClickListener { onOkButtonClick() }
        } else {
            val apiVersion = ApiUtils.getConversationApiVersion(currentUser, intArrayOf(ApiUtils.APIv4, ApiUtils.APIv1))
            ncApi.getRoom(
                ApiUtils.getCredentials(currentUser!!.username, currentUser!!.token),
                ApiUtils.getUrlForRoom(apiVersion, currentUser!!.baseUrl, roomToken)
            )
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .retry(1)
                .subscribe(object : Observer<RoomOverall> {
                    override fun onSubscribe(d: Disposable) {
                        // unused atm
                    }

                    @Suppress("Detekt.LongMethod")
                    override fun onNext(roomOverall: RoomOverall) {
                        conversation = roomOverall.ocs!!.data

                        if (conversation != null && operation === ConversationOperationEnum.OPS_CODE_RENAME_ROOM) {
                            binding?.textEdit?.setText(conversation!!.name)
                        }

                        binding?.textEdit?.setOnEditorActionListener { v, actionId, event ->
                            @Suppress("IMPLICIT_BOXING_IN_IDENTITY_EQUALS")
                            if (actionId === EditorInfo.IME_ACTION_DONE && binding?.okButton?.isEnabled == true) {
                                binding?.okButton?.callOnClick()
                                return@setOnEditorActionListener true
                            }
                            false
                        }

                        textEditAddChangedListener()

                        var labelText = ""
                        when (operation) {
                            ConversationOperationEnum.OPS_CODE_INVITE_USERS,
                            ConversationOperationEnum.OPS_CODE_RENAME_ROOM -> {
                                labelText = resources!!.getString(R.string.nc_call_name)
                                binding?.textEdit?.inputType = InputType.TYPE_CLASS_TEXT
                                binding?.smileyButton?.visibility = View.VISIBLE
                                emojiPopup = binding?.let {
                                    EmojiPopup(
                                        rootView = view,
                                        editText = it.textEdit,
                                        onEmojiPopupShownListener = {
                                            viewThemeUtils.platform.colorImageView(it.smileyButton, ColorRole.PRIMARY)
                                        },
                                        onEmojiPopupDismissListener = {
                                            it.smileyButton.imageTintList = ColorStateList.valueOf(
                                                ResourcesCompat.getColor(
                                                    resources!!,
                                                    R.color.medium_emphasis_text,
                                                    context.theme
                                                )
                                            )
                                        },
                                        onEmojiClickListener = {
                                            binding?.textEdit?.editableText?.append(" ")
                                        }
                                    )
                                }
                            }

                            ConversationOperationEnum.OPS_CODE_JOIN_ROOM -> {
                                // 99 is joining a conversation via password
                                labelText = resources!!.getString(R.string.nc_password)
                                binding?.textEdit?.inputType =
                                    InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
                            }

                            else -> {
                            }
                        }
                        if (PASSWORD_ENTRY_OPERATIONS.contains(operation)) {
                            binding?.textInputLayout?.endIconMode = TextInputLayout.END_ICON_PASSWORD_TOGGLE
                        } else {
                            binding?.textInputLayout?.endIconMode = TextInputLayout.END_ICON_NONE
                        }

                        binding?.textInputLayout?.let { viewThemeUtils.material.colorTextInputLayout(it) }
                        binding?.okButton?.let { viewThemeUtils.material.colorMaterialButtonText(it) }

                        binding?.textInputLayout?.hint = labelText
                        binding?.textInputLayout?.requestFocus()

                        binding?.smileyButton?.setOnClickListener { onSmileyClick() }
                        binding?.okButton?.setOnClickListener { onOkButtonClick() }
                    }

                    override fun onError(e: Throwable) {
                        Log.e("EntryMenuController", "error")
                    }

                    override fun onComplete() {
                        // unused atm
                    }
                })
        }
    }

    private fun textEditAddChangedListener() {
        binding?.textEdit?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
                // unused atm
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                // unused atm
            }

            override fun afterTextChanged(s: Editable) {
                if (!TextUtils.isEmpty(s)) {
                    if (operation === ConversationOperationEnum.OPS_CODE_RENAME_ROOM) {
                        if (conversation!!.name == null || !conversation!!.name.equals(s.toString())) {
                            if (!binding?.okButton?.isEnabled!!) {
                                binding?.okButton?.isEnabled = true
                                binding?.okButton?.alpha = OPACITY_ENABLED
                            }
                            binding?.textInputLayout?.isErrorEnabled = false
                        } else {
                            if (binding?.okButton?.isEnabled == true) {
                                binding?.okButton?.isEnabled = false
                                binding?.okButton?.alpha = OPACITY_DISABLED
                            }
                            binding?.textInputLayout?.error = resources?.getString(R.string.nc_call_name_is_same)
                        }
                    } else if (operation !== ConversationOperationEnum.OPS_CODE_GET_AND_JOIN_ROOM) {
                        if (!binding?.okButton?.isEnabled!!) {
                            binding?.okButton?.isEnabled = true
                            binding?.okButton?.alpha = OPACITY_ENABLED
                        }
                        binding?.textInputLayout?.isErrorEnabled = false
                    } else if (
                        UriUtils.hasHttpProtocollPrefixed(binding?.textEdit?.text.toString()) &&
                        binding?.textEdit?.text.toString().contains("/call/")
                    ) {
                        if (!binding?.okButton?.isEnabled!!) {
                            binding?.okButton?.isEnabled = true
                            binding?.okButton?.alpha = OPACITY_ENABLED
                        }
                        binding?.textInputLayout?.isErrorEnabled = false
                    } else {
                        if (binding?.okButton?.isEnabled == true) {
                            binding?.okButton?.isEnabled = false
                            binding?.okButton?.alpha = OPACITY_DISABLED
                        }
                        binding?.textInputLayout?.error = resources?.getString(R.string.nc_wrong_link)
                    }
                } else {
                    if (binding?.okButton?.isEnabled == true) {
                        binding?.okButton?.isEnabled = false
                        binding?.okButton?.alpha = OPACITY_DISABLED
                    }
                    binding?.textInputLayout?.isErrorEnabled = false
                }
            }
        })
    }

    private fun onSmileyClick() {
        emojiPopup?.toggle()
    }

    private fun onOkButtonClick() {
        if (operation === ConversationOperationEnum.OPS_CODE_JOIN_ROOM) {
            joinRoom()
        } else if (
            operation !== ConversationOperationEnum.OPS_CODE_GET_AND_JOIN_ROOM &&
            operation !== ConversationOperationEnum.OPS_CODE_INVITE_USERS
        ) {
            val bundle = Bundle()
            conversation!!.name = binding?.textEdit?.text.toString()
            bundle.putString(BundleKeys.KEY_ROOM_TOKEN, roomToken)
            bundle.putString(BundleKeys.KEY_NEW_ROOM_NAME, binding?.textEdit?.text.toString())
            bundle.putSerializable(BundleKeys.KEY_OPERATION_CODE, operation)
            router.pushController(
                RouterTransaction.with(OperationsMenuController(bundle))
                    .pushChangeHandler(HorizontalChangeHandler())
                    .popChangeHandler(HorizontalChangeHandler())
            )
        } else if (operation !== ConversationOperationEnum.OPS_CODE_INVITE_USERS) {
            val bundle = Bundle()
            bundle.putSerializable(BundleKeys.KEY_OPERATION_CODE, operation)
            bundle.putString(BundleKeys.KEY_CALL_URL, binding?.textEdit?.text.toString())
            router.pushController(
                RouterTransaction.with(OperationsMenuController(bundle))
                    .pushChangeHandler(HorizontalChangeHandler())
                    .popChangeHandler(HorizontalChangeHandler())
            )
        } else if (operation === ConversationOperationEnum.OPS_CODE_INVITE_USERS) {
            originalBundle.putString(BundleKeys.KEY_CONVERSATION_NAME, binding?.textEdit?.text.toString())
            router.pushController(
                RouterTransaction.with(
                    OperationsMenuController(
                        originalBundle
                    )
                )
                    .pushChangeHandler(HorizontalChangeHandler())
                    .popChangeHandler(HorizontalChangeHandler())
            )
        }
    }

    private fun joinRoom() {
        val bundle = Bundle()
        bundle.putString(BundleKeys.KEY_ROOM_TOKEN, roomToken)
        bundle.putString(BundleKeys.KEY_CONVERSATION_PASSWORD, binding?.textEdit?.text.toString())
        bundle.putSerializable(BundleKeys.KEY_OPERATION_CODE, operation)
        router.pushController(
            RouterTransaction.with(OperationsMenuController(bundle))
                .pushChangeHandler(HorizontalChangeHandler())
                .popChangeHandler(HorizontalChangeHandler())
        )
    }

    init {
        sharedApplication!!.componentApplication.inject(this)

        originalBundle = args
        operation = args.getSerializable(BundleKeys.KEY_OPERATION_CODE) as ConversationOperationEnum
        roomToken = args.getString(KEY_ROOM_TOKEN, "")
        name = args.getString(BundleKeys.KEY_APP_ITEM_NAME, "")
        packageName = args.getString(BundleKeys.KEY_APP_ITEM_PACKAGE_NAME, "")
        // callUrl = args.getString(BundleKeys.KEY_CALL_URL, "")
    }

    companion object {
        private val PASSWORD_ENTRY_OPERATIONS: List<ConversationOperationEnum> =
            immutableListOf(
                ConversationOperationEnum.OPS_CODE_JOIN_ROOM
            )
        const val OPACITY_DISABLED = 0.38f
        const val OPACITY_BUTTON_DISABLED = 0.7f
        const val OPACITY_ENABLED = 1.0f
    }
}
